### PR TITLE
Add configurable option to wait for GitHub external rate limiter

### LINKED ISSUE #48540: GitHub
We need to handle rate limits from GitHub correctly.

### PR DESCRIPTION
**Pull request by pjlast**
Closes #48540

Sets the default GitHub rate limit to infinite, and adds an option to the V3Client to wait based on the rate limit headers returned by GitHub.

**Before this PR**
The default rate limit is set to `5000/3600` when no rate limit default is set, which severely handicaps our interactions with GitHub. It is also a poor approximation for the actual rate limit from GitHub's side.

**After this PR**
Now, if specified, the client will check the provided headers from GItHub and if a rate limit is hit, it will wait the specified amount of time before retrying. This may be quite a long time depending on which rate limit is hit, but it also does not make sense to fail the job when the rate limit is hit and have impartial data.

## Test plan

Tests adjusted.

### CODE REVIEW

File: internal/extsvc/types.go [hunk 1]
```diff
@@ -613,8 +613,8 @@ func GetLimitFromConfig(kind string, config any) (rate.Limit, error) {
 			limit = limitOrInf(c.RateLimit.Enabled, c.RateLimit.RequestsPerHour)
 		}
 	case *schema.GitHubConnection:
-		// 5000 per hour is the default enforced by GitHub on their end
-		limit = rate.Limit(5000.0 / 3600.0)
+		// Use an infinite rate limiter. GitHub has an external rate limiter we obey.
+		limit = rate.Inf
```
Comment (kopancek) [inline internal/extsvc/types.go:617]: Maybe I don't understand everything, but don't we only obey the github rate limiter if `WaitForRateLimit` is `true`? Is it safe to set infinite here for all the cases?

File: enterprise/internal/authz/github/client.go [hunk 2]
```diff
@@ -56,3 +57,7 @@ func (c *ClientAdapter) WithAuthenticator(auther auth.Authenticator) client {
 		V3Client: c.V3Client.WithAuthenticator(auther),
 	}
 }
+
+func (c *ClientAdapter) SetWaitForRateLimit(wait bool) {
+	c.V3Client.WaitForRateLimit = wait
```
Comment (mrnugget) [inline enterprise/internal/authz/github/client.go:62]: Why not make this on by default and only add a "disable wait" method? Seems like the majority of our calls are in background jobs, where waiting makes sense. The exception are calls where we need an immediate response, right? So I'd make the majority the default case and the exception the edge case.

File: internal/extsvc/github/v3.go [hunk 3]
```diff
@@ -204,7 +208,30 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
 	}
 
-	return doRequest(ctx, c.log, c.apiURL, c.auth, c.rateLimitMonitor, c.httpClient, req, result)
+	var resp *httpResponseState
+	resp, err = doRequest(ctx, c.log, c.apiURL, c.auth, c.rateLimitMonitor, c.httpClient, req, result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Retry request after waiting
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:217]: What about these calls here? Shouldn't we remove them? https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+RecommendedWaitForBackgroundOp+file:github%5C.go%24&patternType=standard&sm=0&groupBy=repo

File: internal/extsvc/github/v3.go [hunk 4]
```diff
@@ -204,7 +208,30 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
+	if c.WaitForRateLimit && resp.statusCode == http.StatusForbidden {
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:218]: Could there be other easons for `StatusForbidden` that don't have anything to do with rate limit? Any problems here possibly or are we good with the call to `!known` below?

File: internal/extsvc/github/v3.go [hunk 5]
```diff
@@ -204,7 +208,30 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
+
+		if !known {
+			return resp, err
+		}
+
+		if retry > 0 {
+			time.Sleep(retry)
+			resp, err = doRequest(ctx, c.log, c.apiURL, c.auth, c.rateLimitMonitor, c.httpClient, req, result)
+		} else if remaining == 0 && reset > 0 {
+			time.Sleep(reset)
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:229]: I think these `time.Sleep` should be `SleepWithContext` so we don't block if someone kills the process etc.

File: internal/extsvc/github/v3.go (hunk 4 above)
Comment (pjlast) [reply internal/extsvc/github/v3.go:218]: We should be good with the call to `!known` together with the `retry > 0` and the `remaining == 0 &&  reset > 0` checks.

File: enterprise/internal/authz/github/client.go (hunk 2 above)
Comment (eseliger) [reply enterprise/internal/authz/github/client.go:62]: Agree with Thorsten, all the batch changes calls _do_ want to wait appropriately.

File: internal/extsvc/github/v3.go (hunk 3 above)
Comment (mrnugget) [reply internal/extsvc/github/v3.go:219]: Hmm, but if we take those out, it would still work like before with your new code, right?

It feels weird not knowing when to use what: to I have to wait before sending a request or do I try and then retry if necessary? If possible, I think we should have just one way of doing it. If what you introduce works for this case too, then we should do that.

File: internal/extsvc/github/v3.go [hunk 6]
```diff
@@ -52,12 +53,15 @@ type V3Client struct {
...
 
-	// rateLimitMonitor is the API rate limit monitor.
-	rateLimitMonitor *ratelimit.Monitor
+	// externalRateLimiter is the external API rate limit monitor.
+	externalRateLimiter *ratelimit.Monitor
 
 	// rateLimit is our self-imposed rate limiter
 	rateLimit *ratelimit.InstrumentedLimiter
 
+	// If true, the client will wait and retry a request if external rate limits are encountered
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:62]:
```suggestion
	// WaitForRateLimit determines whether the client will wait and retry a request if external rate limits are encountered.
```

Nitpick, but this is how you do docstrings in Go, ideally.

Also: does this need to be exported, since you have the setter?

File: internal/extsvc/github/v3.go [hunk 7]
```diff
@@ -135,7 +140,7 @@ func (c *V3Client) WithAuthenticator(a auth.Authenticator) *V3Client {
 
 // RateLimitMonitor exposes the rate limit monitor.
 func (c *V3Client) RateLimitMonitor() *ratelimit.Monitor {
-	return c.rateLimitMonitor
+	return c.externalRateLimiter
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:143]: Next PR then renames the method?

File: internal/extsvc/github/v3.go [hunk 8]
```diff
@@ -204,7 +209,38 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
-	return doRequest(ctx, c.log, c.apiURL, c.auth, c.rateLimitMonitor, c.httpClient, req, result)
+	// If rate limit has already been hit, we need to wait before making a request
+	remaining, reset, retry, known := c.externalRateLimiter.Get()
+	if known {
+		if remaining == 0 {
+			timeutil.SleepWithContext(ctx, reset)
+		}
+
+		if retry > 0 {
+			timeutil.SleepWithContext(ctx, retry)
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:220]: Is it now possible to sleep _twice_? Or should these be `if remaining == 0 { ... } else if retry > 0 {`?

File: internal/extsvc/github/v3.go [hunk 9]
```diff
@@ -204,7 +209,38 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
 		return nil, errInternalRateLimitExceeded
 	}
 
-	return doRequest(ctx, c.log, c.apiURL, c.auth, c.rateLimitMonitor, c.httpClient, req, result)
+	// If rate limit has already been hit, we need to wait before making a request
+	remaining, reset, retry, known := c.externalRateLimiter.Get()
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:213]: Why do this here instead of optimistically trying and then sleeping? We don't know whether the state of the rate limiter is up to date, right? So we might as well try and _then_ retry, no?

File: internal/extsvc/github/v3.go [hunk 10]
```diff
@@ -204,7 +209,38 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
+	// If we receive a StatusForbidden, we might have hit a rate limit
+	apiError := &APIError{}
+	if errors.As(err, &apiError) && c.WaitForRateLimit && apiError.Code == http.StatusForbidden {
+		remaining, reset, retry, known := c.externalRateLimiter.Get()
+
+		if retry > 0 {
+			// If retry > 0 then the secondary rate limit has been hit
+			timeutil.SleepWithContext(ctx, retry)
+			resp, err = doRequest(ctx, c.log, c.apiURL, c.auth, c.externalRateLimiter, c.httpClient, req, result)
+		} else if known && remaining == 0 && reset > 0 {
```
Comment (mrnugget) [inline internal/extsvc/github/v3.go:236]: I find these conditions a bit hard to follow: why do we only check `known` in the second branch? I would assume that _both_ would need to be wrapped into `if known`:

```go
if known {
	var sleepTime time.Duration
	if retry > 0 {
		sleepTime = retry
	} else if remaining == 0 && reset > 0 {
		sleepTime = reset
	}
	time.SleepWithContext(ctx, sleepTime)
	resp, err = doRequest(ctx, c.log, c.apiURL, c.auth, c.externalRateLimiter, c.httpClient, req, result)
}

```

File: internal/extsvc/github/v3_test.go [hunk 11]
```diff
@@ -998,6 +1001,77 @@ func TestResponseHasNextPage(t *testing.T) {
...
+		client.WaitForRateLimit = true
+
+		// We do a simple request to test the retry
+		_, err = client.GetVersion(ctx)
+		require.NoError(t, err)
+
+		// We assert that two requests happened
+		assert.True(t, succeeded)
+		assert.False(t, hitSecondaryLimit)
+	})
```
Comment (mrnugget) [inline internal/extsvc/github/v3_test.go:1072]: I'd add a happy case here too

**Review by mrnugget — COMMENTED** [review]
Drive-by comments before hopping into meeting

File: internal/extsvc/github/v3.go (hunk 8 above)
Comment (kopancek) [reply internal/extsvc/github/v3.go:220]: Wouldn't it be easier if the rate limit monitor returned just the retry interval in both cases?

File: internal/extsvc/github/v3.go (hunk 10 above)
Comment (pjlast) [reply internal/extsvc/github/v3.go:236]: I'll add some comments explaining it. But basically: `retry` is the secondary rate limit that can get triggered by GitHub at their discretion. This is just a `Retry-After` header that gets set.

The other three fields are tied together: `known` gets set to true when both `remaining` and `reset` could be determined from the headers. But `known`, `remaining` and `reset` are independent of `retry`

File: internal/extsvc/github/v3.go (hunk 9 above)
Comment (pjlast) [reply internal/extsvc/github/v3.go:213]: If the state of the rate limiter is out of date then `reset` and `retry` will be times in the past, and then the if statements will fail. So we're guaranteed that, if those numbers returned are still positive, then the rate limit is still valid. In the `Monitor` internals these times are stored as absolute dates, and calling `.Get()` turns them into `time.Duration`s

File: internal/extsvc/github/v3.go [hunk 12]
```diff
@@ -204,7 +209,38 @@ func (c *V3Client) request(ctx context.Context, req *http.Request, result any) (
...
+
+	// If we receive a StatusForbidden, we might have hit a rate limit
+	apiError := &APIError{}
+	if errors.As(err, &apiError) && c.WaitForRateLimit && apiError.Code == http.StatusForbidden {
+		remaining, reset, retry, known := c.externalRateLimiter.Get()
+
+		if retry > 0 {
+			// If retry > 0 then the secondary rate limit has been hit
+			timeutil.SleepWithContext(ctx, retry)
+			resp, err = doRequest(ctx, c.log, c.apiURL, c.auth, c.externalRateLimiter, c.httpClient, req, result)
```
Comment (kopancek) [inline internal/extsvc/github/v3.go:235]: Does it mean we only retry once? I would assume we would call `request` here and count the number of allowed retries.

File: internal/repos/github.go [hunk 13]
```diff
@@ -392,18 +389,12 @@ func (s *GitHubSource) ListNamespaces(ctx context.Context, results chan SourceNa
 			return
 		}
 		var pageOrgs []*github.Org
-		pageOrgs, hasNextPage, cost, err = s.v3Client.GetAuthenticatedUserOrgsForPage(ctx, page)
+		pageOrgs, hasNextPage, _, err = s.v3Client.GetAuthenticatedUserOrgsForPage(ctx, page)
```
Comment (kopancek) [inline internal/repos/github.go:392]: In next iteration, should we remove the `cost` from returned parameters of that function? Do we still need it in other places?

File: internal/extsvc/github/v3.go (hunk 8 above)
Comment (pjlast) [reply internal/extsvc/github/v3.go:220]: The retry should be outside of the `known` I'm not sure what went through my head here. Thanks! And yes they should be an `else if`

File: internal/extsvc/github/v3.go (hunk 7 above)
Comment (mrnugget) [reply internal/extsvc/github/v3.go:148]: Could've done it in next PR. Because now you need to update the docstring too 😛

File: internal/extsvc/github/v3.go (hunk 12 above)
Comment (pjlast) [reply internal/extsvc/github/v3.go]: This is a curious point. How many times would you realistically want to retry due to rate limits 🤔 but maybe worth doing anyways, since you could theoretically hit token rate limits and then secondary rate limits.

File: internal/extsvc/github/v3_test.go [hunk 14]
```diff
@@ -998,6 +1002,109 @@ func TestResponseHasNextPage(t *testing.T) {
...
+	srvURL, err := url.Parse(srv.URL)
+	require.NoError(t, err)
+
+	client := NewV3Client(logtest.NoOp(t), "test", srvURL, nil, nil)
+	client.waitForRateLimit = true
+
+	t.Run("primary rate limit hit", func(t *testing.T) {
+		numRequests = 0
+		hitPrimaryLimit = true
+		succeeded = false
```
Comment (mrnugget) [inline internal/extsvc/github/v3_test.go:1049]: Nit: create a `newClient` test helper function in here that creates the test server with the counts, builds the client and returns the client and an assertion function. That way you remove the global vars.

```go
client, done := newClient(t)
_, err = client.GetVersion(ctx)
// ...

done(func(t *testing.T, numRequests int, hitLimit, succeeded bool) {
		assert.True(t, succeeded)
		assert.False(t, hitPrimaryLimit)
		assert.Equal(t, 2, numRequests)
})

**Review by mrnugget — APPROVED** [review]
Cool! Good move on adding that method. I suggest also testing this locally with lots of print-statements to make sure everything really works like it should

**Review by kopancek — APPROVED** [review]
Looks much cleaner to me, thanks!

**Comment by mrnugget** [comment]
btw. what about the GitHub v4 client?

File: internal/ratelimit/monitor.go [hunk 15]
```diff
@@ -159,6 +162,35 @@ func (c *Monitor) RecommendedWaitForBackgroundOp(cost int) (timeRemaining time.D
...
+// and sleeps an amount of time recommended by the external rate limiter.
+// It returns true if rate limiting was applying, and false if not.
+// This can be used to determine whether or not a request should be retried.
+func (c *Monitor) WaitForRateLimit(ctx context.Context) bool {
+	// If the "Retry-After" header was set, we need to wait the specified amount of time.
+	if !c.retry.IsZero() {
+		if remaining := c.retry.Sub(c.now()); remaining > 0 {
+			timeutil.SleepWithContext(ctx, remaining)
+			return true
+		}
```
Comment (eseliger) [inline internal/ratelimit/monitor.go:175]: what the go rate limiter does, and I think we would want here, too is that it also checks the context deadline. If the time by which we could run the request is post-deadline, we can fail immediately instead of sleeping for long.

File: internal/repos/github.go [hunk 16]
```diff
@@ -180,9 +179,9 @@ func newGithubSource(
 	}
 
 	for resource, monitor := range map[string]*ratelimit.Monitor{
-		"rest":    v3Client.RateLimitMonitor(),
+		"rest":    v3Client.ExternalRateLimiter(),
 		"graphql": v4Client.RateLimitMonitor(),
```
Comment (eseliger) [inline internal/repos/github.go:183]: any chance we could get this for v4, too or is this not required? Asking for a batch changes friend.

File: enterprise/internal/authz/github/client.go [hunk 17]
```diff
@@ -42,6 +42,7 @@ type client interface {
 
 	GetAuthenticatedOAuthScopes(ctx context.Context) ([]string, error)
 	WithAuthenticator(auther auth.Authenticator) client
+	SetWaitForRateLimit(wait bool)
```
Comment (eseliger) [inline enterprise/internal/authz/github/client.go:45]: this seems unused from the diff 🤔

**Review by eseliger — COMMENTED** [review]
nice work :)
